src/main/java/com/mojang/serialization/DataResult.java

```diff
@@ -20,298 +18,354 @@
+            final Lifecycle combinedLifecycle = lifecycle.add(functionResult.lifecycle());
+            if (functionResult instanceof final Success<Function<R, R2>> funcSuccess) {
+                return new Success<>(funcSuccess.value.apply(value), combinedLifecycle);
+            } else if (functionResult instanceof final Error<Function<R, R2>> funcError) {
+                return new Error<>(funcError.messageSupplier, funcError.partialValue.map(f -> f.apply(value)), combinedLifecycle);
+            } else {
+                throw new UnsupportedOperationException();
             }
-            return new PartialResult<>(message, Optional.empty());
         }
 
+        @Override
+        public DataResult<R> setPartial(final Supplier<R> partial) {
+            return this;
+        }
+
+        @Override
+        public DataResult<R> setPartial(final R partial) {
+            return this;
+        }
+
+        @Override
+        public DataResult<R> mapError(final UnaryOperator<String> function) {
+            return this;
+        }
+
+        @Override
+        public DataResult<R> setLifecycle(final Lifecycle lifecycle) {
+            if (this.lifecycle.equals(lifecycle)) {
+                return this;
+            }
+            return new Success<>(value, lifecycle);
+        }
+
+        @Override
+        public String toString() {
+            return "DataResult.Success[" + value + "]";
+        }
+    }
+
+    record Error<R>(
+        Supplier<String> messageSupplier,
+        Optional<R> partialValue,
+        Lifecycle lifecycle
+    ) implements DataResult<R> {
         public String message() {
-            return message.get();
+            return messageSupplier.get();
+        }
+
+        @Override
+        public Optional<R> result() {
+            return Optional.empty();
+        }
+
+        @Override
+        public Optional<Error<R>> error() {
+            return Optional.of(this);
+        }
+
+        @Override
+        public Optional<R> resultOrPartial(final Consumer<String> onError) {
+            onError.accept(messageSupplier.get());
+            return partialValue;
+        }
+
+        @Override
+        public Optional<R> resultOrPartial() {
+            return partialValue;
+        }
+
+        @Override
+        public <E extends Throwable> R getOrThrow(final Function<String, E> exceptionSupplier) throws E {
+            throw exceptionSupplier.apply(message());
+        }
+
+        @Override
+        public <E extends Throwable> R getPartialOrThrow(final Function<String, E> exceptionSupplier) throws E {
+            if (partialValue.isPresent()) {
+                return partialValue.get();
+            }
+            throw exceptionSupplier.apply(message());
+        }
+
+        @Override
+        public <T> Error<T> map(final Function<? super R, ? extends T> function) {
+            if (partialValue.isEmpty()) {
+                // noinspection unchecked
```

Comment:
Hmm, for consistent style it should be rather `@SuppressWarnings("unchecked")` (even though it ends up covering whole method)